Using decorator doesn't count as usage by eslint

Have the following code:
```
import autobind from 'autobind-decorator';

class TextareaAutosize {
  ...

  @autobind
  _onChange(e) { ... }
}
```
which results into:
```
   7:7   error    autobind is defined but never used                                                no-unused-vars
```

--- a/index.js
+++ b/index.js
@@ -54,6 +54,41 @@ function monkeypatch() {
     estraverse.VisitorKeys.TypeAlias = TypeAliasKeys;
     return results;
   };
+
+  // monkeypatch escope/referencer
+  var referencerLoc;
+  try {
+    var referencerLoc = Module._resolveFilename("./referencer", escopeMod);
+  } catch (err) {
+    throw new ReferenceError("couldn't resolve escope/referencer");
+  }
+  var referencer = require(referencerLoc);
+
+  // if there are decotators, then visit each
+  function visitDecorators(node) {
+    if (!node.decorators) {
+      return;
+    }
+    for (var i = 0; i < node.decorators.length; i++) {
+      if (node.decorators[i].expression) {
+        this.visit(node.decorators[i]);
+      }
+    }
+  }
+
+  // monkeypatch referencer methods to visit decorators
+  var visitClass = referencer.prototype.visitClass;
+  referencer.prototype.visitClass = function (node) {
+    // visit decorators that are in: Class Declaration
+    visitDecorators.call(this, node);
+    visitClass.call(this, node);
+  }
+  var visitProperty = referencer.prototype.visitProperty;
+  referencer.prototype.visitProperty = function (node) {
+    // visit decorators that are in: Visit Property / MethodDefinition
+    visitDecorators.call(this, node);
+    visitProperty.call(this, node);
+  }
 }
 
 exports.attachComments = function (ast, comments, tokens) {

--- a/test/non-regression.js
+++ b/test/non-regression.js
@@ -55,7 +55,7 @@ describe("verify", function () {
     verifyAndAssertMessages(
       "import Foo from 'foo';\n" +
       "export default Foo;",
-      { },
+      {},
       []
     );
   });
@@ -162,5 +162,97 @@ describe("verify", function () {
       { "comma-spacing": 1 },
       []
     );
+  });
+
+  describe("decorators #72", function () {
+    it("class declaration", function () {
+      verifyAndAssertMessages(
+        [
+          "import classDeclaration from 'decorator';",
+          "import decoratorParameter from 'decorator';",
+          "@classDeclaration(decoratorParameter)",
+          "@classDeclaration",
+          "class TextareaAutosize {}"
+        ].join("\n"),
+        { "no-unused-vars": 1 },
+        []
+      );
+    });
+
+    it("method definition", function () {
+      verifyAndAssertMessages(
+        [
+          "import classMethodDeclarationA from 'decorator';",
+          "import decoratorParameter from 'decorator';",
+          "class TextareaAutosize {",
+            "@classMethodDeclarationA(decoratorParameter)",
+            "@classMethodDeclarationA",
+            "methodDeclaration(e) {",
+              "e();",
+            "}",
+          "}"
+        ].join("\n"),
+        { "no-unused-vars": 1 },
+        []
+      );
+    });
+
+    it("method definition get/set", function () {
+      verifyAndAssertMessages(
+        [
+          "import classMethodDeclarationA from 'decorator';",
+          "import decoratorParameter from 'decorator';",
+          "class TextareaAutosize {",
+            "@classMethodDeclarationA(decoratorParameter)",
+            "@classMethodDeclarationA",
+            "get bar() { }",
+            "@classMethodDeclarationA(decoratorParameter)",
+            "@classMethodDeclarationA",
+            "set bar() { }",
+          "}"
+        ].join("\n"),
+        { "no-unused-vars": 1 },
+        []
+      );
+    });
+
+    it("object property", function () {
+      verifyAndAssertMessages(
+        [
+          "import classMethodDeclarationA from 'decorator';",
+          "import decoratorParameter from 'decorator';",
+          "var obj = {",
+            "@classMethodDeclarationA(decoratorParameter)",
+            "@classMethodDeclarationA",
+            "methodDeclaration(e) {",
+              "e();",
+            "}",
+          "};",
+          "obj;"
+        ].join("\n"),
+        { "no-unused-vars": 1 },
+        []
+      );
+    });
+
+    it("object property get/set", function () {
+      verifyAndAssertMessages(
+        [
+          "import classMethodDeclarationA from 'decorator';",
+          "import decoratorParameter from 'decorator';",
+          "var obj = {",
+            "@classMethodDeclarationA(decoratorParameter)",
+            "@classMethodDeclarationA",
+            "get bar() { },",
+            "@classMethodDeclarationA(decoratorParameter)",
+            "@classMethodDeclarationA",
+            "set bar() { }",
+          "};",
+          "obj;"
+        ].join("\n"),
+        { "no-unused-vars": 1 },
+        []
+      );
+    });
   });
 });

